From 558c545764a285ece1764e22497e6770b594183a Mon Sep 17 00:00:00 2001
From: Stefan Breunig <stefan-ffmpeg-devel@breunig.xyz>
Date: Sun, 3 May 2026 07:59:31 +0200
Subject: [PATCH] avfilter/vf_frei0r: support mixer2 and mixer3 plugins

https://github.com/breunigs/ffmpeg

Using dynamic inputs to support the additional plugins seemed
preferable over separate filters. For backwards compatibility, all
frame properties are copied from the first input, adjusting only
the necessary ones for multiple inputs.

Signed-off-by: Stefan Breunig <stefan-ffmpeg-devel@breunig.xyz>
---
diff --git a/doc/filters.texi b/doc/filters.texi
index 2cae41c7c5..abb4ab34ea 100644
--- a/doc/filters.texi
+++ b/doc/filters.texi
@@ -14934,8 +14934,19 @@ positions:
 @example
 frei0r=perspective:0.2/0.2|0.8/0.2
 @end example
+
+@item
+Apply the addition effect, which takes two inputs.
+@example
+ffmpeg -i INPUT1 -i INPUT2 -filter_complex "[0][1]frei0r=addition" OUTPUT
+@end example
 @end itemize
 
+If multiple inputs are given, they all must have the same width and height.
+Properties are taken from the first input.
+
+This filter also supports the @ref{framesync} options.
+
 For more information, see
 @url{http://frei0r.dyne.org}
 
diff --git a/libavfilter/Makefile b/libavfilter/Makefile
index 03bf51d3fd..861c80482a 100644
--- a/libavfilter/Makefile
+++ b/libavfilter/Makefile
@@ -329,7 +329,7 @@ OBJS-$(CONFIG_FRAMERATE_FILTER)              += vf_framerate.o
 OBJS-$(CONFIG_FRAMESTEP_FILTER)              += vf_framestep.o
 OBJS-$(CONFIG_FREEZEDETECT_FILTER)           += vf_freezedetect.o
 OBJS-$(CONFIG_FREEZEFRAMES_FILTER)           += vf_freezeframes.o
-OBJS-$(CONFIG_FREI0R_FILTER)                 += vf_frei0r.o
+OBJS-$(CONFIG_FREI0R_FILTER)                 += vf_frei0r.o framesync.o
 OBJS-$(CONFIG_FSPP_FILTER)                   += vf_fspp.o vf_fsppdsp.o qp_table.o
 OBJS-$(CONFIG_FSYNC_FILTER)                  += vf_fsync.o
 OBJS-$(CONFIG_GBLUR_FILTER)                  += vf_gblur.o
diff --git a/libavfilter/vf_frei0r.c b/libavfilter/vf_frei0r.c
index 722d6dca92..a3bbd2ba43 100644
--- a/libavfilter/vf_frei0r.c
+++ b/libavfilter/vf_frei0r.c
@@ -41,6 +41,7 @@
 #include "avfilter.h"
 #include "filters.h"
 #include "formats.h"
+#include "framesync.h"
 #include "video.h"
 
 #ifdef __APPLE__
@@ -64,6 +65,7 @@ typedef void (*f0r_get_param_value_f)(f0r_instance_t instance, f0r_param_t param
 typedef struct Frei0rContext {
     const AVClass *class;
     f0r_update_f update;
+    f0r_update2_f update2;
     void *dl_handle;            /* dynamic library handle   */
     f0r_instance_t instance;
     f0r_plugin_info_t plugin_info;
@@ -75,6 +77,9 @@ typedef struct Frei0rContext {
     f0r_destruct_f        destruct;
     f0r_deinit_f          deinit;
 
+    FFFrameSync fs;
+    AVFrame **frames;
+    int nb_inputs;
     char *dl_name;
     char *params;
     AVRational framerate;
@@ -189,8 +194,8 @@ static int load_path(AVFilterContext *ctx, void **handle_ptr, const char *prefix
     return 0;
 }
 
-static av_cold int frei0r_init(AVFilterContext *ctx,
-                               const char *dl_name, int type)
+static av_cold int frei0r_init(AVFilterContext *ctx, const char *dl_name,
+                               int min_inputs, int max_inputs)
 {
     Frei0rContext *s = ctx->priv;
     f0r_init_f            f0r_init;
@@ -268,7 +273,6 @@ static av_cold int frei0r_init(AVFilterContext *ctx,
         !(s->get_param_info  = load_sym(ctx, "f0r_get_param_info" )) ||
         !(s->get_param_value = load_sym(ctx, "f0r_get_param_value")) ||
         !(s->set_param_value = load_sym(ctx, "f0r_set_param_value")) ||
-        !(s->update          = load_sym(ctx, "f0r_update"         )) ||
         !(s->construct       = load_sym(ctx, "f0r_construct"      )) ||
         !(s->destruct        = load_sym(ctx, "f0r_destruct"       )) ||
         !(s->deinit          = load_sym(ctx, "f0r_deinit"         )))
@@ -281,7 +285,23 @@ static av_cold int frei0r_init(AVFilterContext *ctx,
 
     f0r_get_plugin_info(&s->plugin_info);
     pi = &s->plugin_info;
-    if (pi->plugin_type != type) {
+
+    s->nb_inputs = pi->plugin_type == F0R_PLUGIN_TYPE_SOURCE ? 0 :
+                   pi->plugin_type == F0R_PLUGIN_TYPE_FILTER ? 1 :
+                   pi->plugin_type == F0R_PLUGIN_TYPE_MIXER2 ? 2 :
+                   pi->plugin_type == F0R_PLUGIN_TYPE_MIXER3 ? 3 : -1;
+
+    if (s->nb_inputs < 2) {
+        s->update = load_sym(ctx, "f0r_update");
+        if (!s->update)
+            return AVERROR(EINVAL);
+    } else {
+        s->update2 = load_sym(ctx, "f0r_update2");
+        if (!s->update2)
+            return AVERROR(EINVAL);
+    }
+
+    if (s->nb_inputs < min_inputs || s->nb_inputs > max_inputs) {
         av_log(ctx, AV_LOG_ERROR,
                "Invalid type '%s' for this plugin\n",
                pi->plugin_type == F0R_PLUGIN_TYPE_FILTER ? "filter" :
@@ -304,13 +324,6 @@ static av_cold int frei0r_init(AVFilterContext *ctx,
     return 0;
 }
 
-static av_cold int filter_init(AVFilterContext *ctx)
-{
-    Frei0rContext *s = ctx->priv;
-
-    return frei0r_init(ctx, s->dl_name, F0R_PLUGIN_TYPE_FILTER);
-}
-
 static av_cold void uninit(AVFilterContext *ctx)
 {
     Frei0rContext *s = ctx->priv;
@@ -321,21 +334,154 @@ static av_cold void uninit(AVFilterContext *ctx)
         s->deinit();
     if (s->dl_handle)
         dlclose(s->dl_handle);
+
+    ff_framesync_uninit(&s->fs);
+    av_freep(&s->frames);
 }
 
-static int config_input_props(AVFilterLink *inlink)
+static int filter_frame(FFFrameSync *fs)
 {
-    AVFilterContext *ctx = inlink->dst;
+    AVFilterContext *ctx = fs->parent;
+    AVFilterLink *outlink = ctx->outputs[0];
+    FilterLink *ol = ff_filter_link(outlink);
+    Frei0rContext *s = fs->opaque;
+    AVFrame **in = s->frames;
+    AVFrame *aligned[3] = { NULL };
+    AVFrame *out;
+    int i, ret;
+    double time;
+
+    for (i = 0; i < s->nb_inputs; i++) {
+        ret = ff_framesync_get_frame(&s->fs, i, &in[i], 0);
+        if (ret < 0)
+            return ret;
+    }
+
+    if (ctx->is_disabled) {
+        out = av_frame_clone(in[0]);
+        if (!out)
+            return AVERROR(ENOMEM);
+        return ff_filter_frame(outlink, out);
+    }
+
+    /* align parameter is the line alignment, not the buffer alignment.
+     * frei0r expects line size to be width*4 so we want an align of 1
+     * to ensure lines aren't padded out. */
+    out = ff_default_get_video_buffer2(outlink, outlink->w, outlink->h, 1);
+    if (!out)
+        return AVERROR(ENOMEM);
+
+    ret = av_frame_copy_props(out, in[0]);
+    if (ret < 0) {
+        av_frame_free(&out);
+        return ret;
+    }
+    out->pts = av_rescale_q(s->fs.pts, s->fs.time_base, outlink->time_base);
+    if (s->nb_inputs >= 2 && ol->frame_rate.den == 0)
+        out->duration = 0;
+
+    for (i = 0; i < s->nb_inputs; i++) {
+        if (in[i]->linesize[0] != out->linesize[0]) {
+            aligned[i] = ff_default_get_video_buffer2(outlink, outlink->w, outlink->h, 1);
+            if (!aligned[i]) {
+                ret = AVERROR(ENOMEM);
+                goto end;
+            }
+            ret = av_frame_copy(aligned[i], in[i]);
+            if (ret < 0)
+                goto end;
+            in[i] = aligned[i];
+        }
+    }
+
+    time = s->fs.pts * av_q2d(s->fs.time_base);
+    if (s->nb_inputs == 1) {
+        s->update(s->instance, time,
+                  (const uint32_t *)in[0]->data[0],
+                  (uint32_t *)out->data[0]);
+    } else {
+        s->update2(s->instance, time,
+                   (const uint32_t *)in[0]->data[0],
+                   (const uint32_t *)in[1]->data[0],
+                   s->nb_inputs >= 3 ? (const uint32_t *)in[2]->data[0] : NULL,
+                   (uint32_t *)out->data[0]);
+    }
+
+    ret = ff_filter_frame(outlink, out);
+
+end:
+    for (i = 0; i < s->nb_inputs; i++)
+        av_frame_free(&aligned[i]);
+
+    return ret;
+}
+
+static int config_link_props(AVFilterLink *outlink)
+{
+    AVFilterContext *ctx = outlink->src;
     Frei0rContext *s = ctx->priv;
+    FilterLink *il = ff_filter_link(ctx->inputs[0]);
+    FilterLink *ol = ff_filter_link(outlink);
+    int width = ctx->inputs[0]->w;
+    int height = ctx->inputs[0]->h;
+    int i, ret;
+
+    for (i = 1; i < s->nb_inputs; i++) {
+        if (ctx->inputs[i]->w != width || ctx->inputs[i]->h != height) {
+            av_log(ctx, AV_LOG_ERROR,
+                   "Input %d dimensions %dx%d do not match input %d dimensions %dx%d.\n",
+                   i, ctx->inputs[i]->w, ctx->inputs[i]->h, 0, width, height);
+            return AVERROR(EINVAL);
+        }
+    }
 
     if (s->destruct && s->instance)
         s->destruct(s->instance);
-    if (!(s->instance = s->construct(inlink->w, inlink->h))) {
+    if (!(s->instance = s->construct(width, height))) {
         av_log(ctx, AV_LOG_ERROR, "Impossible to load frei0r instance.\n");
         return AVERROR(EINVAL);
     }
 
-    return set_params(ctx, s->params);
+    ret = set_params(ctx, s->params);
+    if (ret < 0)
+        return ret;
+
+    outlink->w = width;
+    outlink->h = height;
+    outlink->sample_aspect_ratio = ctx->inputs[0]->sample_aspect_ratio;
+    ol->frame_rate = il->frame_rate;
+
+    for (i = 1; i < s->nb_inputs; i++) {
+        il = ff_filter_link(ctx->inputs[i]);
+        if (ol->frame_rate.num != il->frame_rate.num ||
+            ol->frame_rate.den != il->frame_rate.den) {
+            av_log(ctx, AV_LOG_VERBOSE,
+                    "Video inputs have different frame rates, output will be VFR\n");
+            ol->frame_rate = av_make_q(1, 0);
+            break;
+        }
+    }
+
+    ret = ff_framesync_init(&s->fs, ctx, s->nb_inputs);
+    if (ret < 0)
+        return ret;
+
+    s->fs.opaque = s;
+    s->fs.on_event = filter_frame;
+
+    for (i = 0; i < s->nb_inputs; i++) {
+        AVFilterLink *inlink = ctx->inputs[i];
+
+        s->fs.in[i].time_base = inlink->time_base;
+        s->fs.in[i].sync      = 1;
+        s->fs.in[i].before    = EXT_STOP;
+        s->fs.in[i].after     = EXT_INFINITY;
+    }
+
+    ret = ff_framesync_configure(&s->fs);
+    outlink->time_base = s->fs.time_base;
+
+    return ret;
 }
 
 static int query_formats(const AVFilterContext *ctx,
@@ -365,43 +511,40 @@ static int query_formats(const AVFilterContext *ctx,
     return ff_set_common_formats2(ctx, cfg_in, cfg_out, formats);
 }
 
-static int filter_frame(AVFilterLink *inlink, AVFrame *in)
+static int filter_activate(AVFilterContext *ctx)
 {
-    Frei0rContext *s = inlink->dst->priv;
-    AVFilterLink *outlink = inlink->dst->outputs[0];
-    /* align parameter is the line alignment, not the buffer alignment.
-     * frei0r expects line size to be width*4 so we want an align of 1
-     * to ensure lines aren't padded out. */
-    AVFrame *out = ff_default_get_video_buffer2(outlink, outlink->w, outlink->h, 1);
-    if (!out)
-        goto fail;
+    Frei0rContext *s = ctx->priv;
+    return ff_framesync_activate(&s->fs);
+}
 
-    av_frame_copy_props(out, in);
+static av_cold int filter_init(AVFilterContext *ctx)
+{
+    Frei0rContext *s = ctx->priv;
+    int i, ret;
 
-    if (in->linesize[0] != out->linesize[0]) {
-        AVFrame *in2 = ff_default_get_video_buffer2(outlink, outlink->w, outlink->h, 1);
-        if (!in2)
-            goto fail;
-        av_frame_copy(in2, in);
-        if (av_frame_copy_props(in2, in) < 0) {
-            av_frame_free(&in2);
-            goto fail;
-        }
-        av_frame_free(&in);
-        in = in2;
-    }
+    ret = frei0r_init(ctx, s->dl_name, 1, 3);
+    if (ret < 0)
+        return ret;
 
-    s->update(s->instance, in->pts * av_q2d(inlink->time_base),
-                   (const uint32_t *)in->data[0],
-                   (uint32_t *)out->data[0]);
+    s->frames = av_calloc(s->nb_inputs, sizeof(*s->frames));
+    if (!s->frames)
+        return AVERROR(ENOMEM);
 
-    av_frame_free(&in);
+    for (i = 0; i < s->nb_inputs; i++) {
+        AVFilterPad pad = {
+            .type = AVMEDIA_TYPE_VIDEO,
+            .name = av_asprintf("input%d", i),
+        };
 
-    return ff_filter_frame(outlink, out);
-fail:
-    av_frame_free(&in);
-    av_frame_free(&out);
-    return AVERROR(ENOMEM);
+        if (!pad.name)
+            return AVERROR(ENOMEM);
+
+        ret = ff_append_inpad_free_name(ctx, &pad);
+        if (ret < 0)
+            return ret;
+    }
+
+    return 0;
 }
 
 static int process_command(AVFilterContext *ctx, const char *cmd, const char *args,
@@ -426,14 +569,13 @@ static const AVOption frei0r_options[] = {
     { NULL }
 };
 
-AVFILTER_DEFINE_CLASS(frei0r);
+FRAMESYNC_DEFINE_CLASS(frei0r, Frei0rContext, fs);
 
-static const AVFilterPad avfilter_vf_frei0r_inputs[] = {
+static const AVFilterPad avfilter_vf_frei0r_outputs[] = {
     {
         .name         = "default",
         .type         = AVMEDIA_TYPE_VIDEO,
-        .config_props = config_input_props,
-        .filter_frame = filter_frame,
+        .config_props = config_link_props,
     },
 };
 
@@ -441,12 +583,13 @@ const FFFilter ff_vf_frei0r = {
     .p.name        = "frei0r",
     .p.description = NULL_IF_CONFIG_SMALL("Apply a frei0r effect."),
     .p.priv_class  = &frei0r_class,
-    .p.flags       = AVFILTER_FLAG_SUPPORT_TIMELINE_GENERIC,
+    .p.flags       = AVFILTER_FLAG_DYNAMIC_INPUTS | AVFILTER_FLAG_SUPPORT_TIMELINE_GENERIC,
+    .preinit       = frei0r_framesync_preinit,
     .init          = filter_init,
     .uninit        = uninit,
+    .activate      = filter_activate,
     .priv_size     = sizeof(Frei0rContext),
-    FILTER_INPUTS(avfilter_vf_frei0r_inputs),
-    FILTER_OUTPUTS(ff_video_default_filterpad),
+    FILTER_OUTPUTS(avfilter_vf_frei0r_outputs),
     FILTER_QUERY_FUNC2(query_formats),
     .process_command = process_command,
 };
@@ -458,7 +601,7 @@ static av_cold int source_init(AVFilterContext *ctx)
     s->time_base.num = s->framerate.den;
     s->time_base.den = s->framerate.num;
 
-    return frei0r_init(ctx, s->dl_name, F0R_PLUGIN_TYPE_SOURCE);
+    return frei0r_init(ctx, s->dl_name, 0, 0);
 }
 
 static int source_config_props(AVFilterLink *outlink)
